### DISCUSSION TITLE
Support bash pipes in custom commands

### DISCUSSION THREAD

**Discussion by mludv** [explosion/spaCy#6964] [Help: Coding & Implementations] [answered]
Thanks for a great 3.0 release!

I was working on converting some Makefile-based corpus-fetching scripts to the new Spacy project format and noticed that the "script" section doesn't support bash features like pipes and stdout redirects.

In my case I'm using `iconv` which always write to stdout, and I want to redirect to an output file instead.

From a quick look in the source code I assume this is because you are doing some preprocessing on the command  provided. I solved the problem by wrapping everything in `bash -c '<command>'`, but I'm thinking maybe other users will encounter the same issue.

Is there a solution or should the workaround perhaps be documented somewhere?

**Comment by honnibal** [comment, marked as answer]
Yeah, thanks for pointing this out. I'll think about the best way to explain this in the docs.

Having pipes or other bash features in the commands could be good for some situations, but we decided that we don't want to call `shell=True` when we trigger the subprocess, to reduce the layers of execution and the number of things that can go wrong. Using `bash -c` as you have is one workaround. Another workaround would be to put the code in a separate script, and invoke it from the project file. That solution does make it more difficult to interpolate into the command though.
